web/metrics_controller_test.go

```diff
@@ -0,0 +1,113 @@
+package web_test
+
+import (
+	"github.com/manyminds/api2go/jsonapi"
+	"github.com/smartcontractkit/chainlink/internal/cltest"
+	"github.com/smartcontractkit/chainlink/store/models"
+	"github.com/smartcontractkit/chainlink/web"
+	"github.com/stretchr/testify/assert"
+	"go.uber.org/multierr"
+	"testing"
+)
+
+func BenchmarkMetricsController_Index(b *testing.B) {
+	app, cleanup := cltest.NewApplicationWithKeyStore()
+	defer cleanup()
+	client := app.NewHTTPClient()
+	setupStatsControllerIndex(app)
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		headers := map[string]string{"Authorization": "Bearer " + app.Config.MetricsBearerToken}
+		resp, cleanup := client.Get("/v2/metrics", headers)
+		defer cleanup()
+		assert.Equal(b, 200, resp.StatusCode, "Response should be successful")
+	}
+}
+
+func TestMetricsController_Index(t *testing.T) {
+	t.Parallel()
+
+	app, cleanup := cltest.NewApplicationWithKeyStore()
+	defer cleanup()
+	client := app.NewHTTPClient()
+
+	headers := map[string]string{"Authorization": "Bearer " + app.Config.MetricsBearerToken}
+
+	j1, j2, err := setupStatsControllerIndex(app)
+	assert.NoError(t, err)
+
+	resp, cleanup := client.Get("/v2/metrics?size=x", headers)
+	defer cleanup()
+	cltest.AssertServerResponse(t, resp, 422)
+
+	resp, cleanup = client.Get("/v2/metrics?size=1", headers)
+	defer cleanup()
+	cltest.AssertServerResponse(t, resp, 200)
+	body := cltest.ParseResponseBody(resp)
+
+	metaCount, err := cltest.ParseJSONAPIResponseMetaCount(body)
+	assert.NoError(t, err)
+	assert.Equal(t, 2, metaCount)
+
+	var links jsonapi.Links
+	stats := models.JobSpecMetrics{}
+	err = web.ParsePaginatedResponse(body, &stats, &links)
+	assert.NoError(t, err)
+	assert.NotEmpty(t, links["next"].Href)
+	assert.Empty(t, links["prev"].Href)
+
+	assert.Len(t, stats.JobSpecCounts, 1)
+	assert.Equal(t, stats.JobSpecCounts[0].AdaptorCount["noop"], 1, "Should have noop as an adaptor")
+	assert.Equal(t, j1.ID, stats.JobSpecCounts[0].ID, "should have the same ID")
+
+	resp, cleanup = client.Get(links["next"].Href, headers)
+	defer cleanup()
+	cltest.AssertServerResponse(t, resp, 200)
+
+	stats = models.JobSpecMetrics{}
+	err = web.ParsePaginatedResponse(cltest.ParseResponseBody(resp), &stats, &links)
+	assert.NoError(t, err)
+	assert.Empty(t, links["next"])
+	assert.NotEmpty(t, links["prev"])
+
+	assert.Len(t, stats.JobSpecCounts, 1)
+	assert.Equal(t, stats.JobSpecCounts[0].RunCount, 1, "Should have a single run")
+	assert.Equal(t, stats.JobSpecCounts[0].StatusCount["completed"], 1, "Should have a single completed run")
+	assert.Equal(t, stats.JobSpecCounts[0].AdaptorCount["noop"], 1, "Should have noop as an adaptor")
+	assert.Equal(t, stats.JobSpecCounts[0].ParamCount["url"][0].Value, "https://chain.link", "Should include the same URL")
+	assert.Equal(t, stats.JobSpecCounts[0].ParamCount["url"][0].Count, 1, "Should include a url")
```

Comment:
A lot of this seems like testing of logic that is already tested in the `service` package. You could get away with just testing some minimal piece of it here, focus on the controller specific parts here, and then you don't have to update it in two places when you make a change.